components/automate-cli/cmd/chef-automate/cmdUtils.go

```diff
@@ -0,0 +1,446 @@
+	}
+
+	if err != nil {
+		return nodeIps, err
+	}
+	if len(nodeIps) == 0 {
+		errMsg := fmt.Sprintf("No %s IPs are found", remoteService)
+		return nodeIps, errors.New(errMsg)
+	}
+	return nodeIps, nil
+}
+
+// getSshDetails will return the SSH details.
+func getSshDetails(infra *AutomteHAInfraDetails) *SSHConfig {
+	sshConfig := &SSHConfig{
+		sshUser:    infra.Outputs.SSHUser.Value,
+		sshPort:    infra.Outputs.SSHPort.Value,
+		sshKeyFile: infra.Outputs.SSHKeyFile.Value,
+	}
+	return sshConfig
+}
+
+// getFrontendIPs will return the Frontend Node Ips.
+func getFrontendIPs(ip string, infra *AutomteHAInfraDetails) ([]string, error) {
+	frontendIps := append(infra.Outputs.AutomatePrivateIps.Value, infra.Outputs.ChefServerPrivateIps.Value...)
+	if ip != "" {
+		if isValidIP(ip, frontendIps) {
+			return []string{ip}, nil
+		}
+		return []string{}, errors.New("Please Enter Valid frontend IP")
+	}
+	return frontendIps, nil
+
+}
+
+// getAutomateIPs will return the Automate Node Ips.
+func getAutomateIPs(single bool, ip string, infra *AutomteHAInfraDetails) ([]string, error) {
+	automateIps := infra.Outputs.AutomatePrivateIps.Value
+	if ip != "" {
+		if isValidIP(ip, automateIps) {
+			return []string{ip}, nil
+		}
+		return []string{}, errors.New("Please Enter Valid automate IP")
+	} else if single {
+		ip, err := GetSingleIp(automateIps)
+		return []string{ip}, err
+	}
+	return automateIps, nil
+}
+
+// getChefserverIPs will return the Chefserver Node Ips.
+func getChefserverIPs(single bool, ip string, infra *AutomteHAInfraDetails) ([]string, error) {
+	chefServerIps := infra.Outputs.ChefServerPrivateIps.Value
+	if ip != "" {
+		if isValidIP(ip, chefServerIps) {
+			return []string{ip}, nil
+		}
+		return []string{}, errors.New("Please Enter Valid chef-server IP")
+	} else if single {
+		ip, err := GetSingleIp(chefServerIps)
+		return []string{ip}, err
+	}
+	return chefServerIps, nil
+
+}
+
+// getPostgresqlIPs will return the Postgresql Node Ips.
+func getPostgresqlIPs(single bool, ip string, infra *AutomteHAInfraDetails) ([]string, error) {
+	postgresqlIps := infra.Outputs.PostgresqlPrivateIps.Value
+	if ip != "" {
+		if isValidIP(ip, postgresqlIps) {
+			return []string{ip}, nil
+		}
+		return []string{}, errors.New("Please Enter Valid postgresql IP")
+	} else if single {
+		ip, err := GetSingleIp(postgresqlIps)
+		return []string{ip}, err
+	}
+	return postgresqlIps, nil
+
+}
+
+// getOpensearchIPs will return the Opensearch Node Ips.
+func getOpensearchIPs(single bool, ip string, infra *AutomteHAInfraDetails) ([]string, error) {
+	opensearchIps := infra.Outputs.OpensearchPrivateIps.Value
+	if ip != "" {
+		if isValidIP(ip, opensearchIps) {
+			return []string{ip}, nil
+		}
+		return []string{}, errors.New("Please Enter Valid opensearch IP")
+	} else if single {
+		ip, err := GetSingleIp(opensearchIps)
+		return []string{ip}, err
+	}
+	return opensearchIps, nil
+}
+
```

Comment:
getOpensearchIPs, getPostgresqlIPs, getChefserverIPs, getAutomateIPs, and getFrontendIP all of these functions have exactly the same logic for a different set of IPs, we can merge all of them in a single function to avoid repetitive codes, something like below...

`
func getNodeIPs(single bool, ip string, infra *AutomteHAInfraDetails, type string) ([]string, error) {
     nodeIps := {}
      switch {
      case type == CONST_OPENSEARCH:
	  nodeIps := infra.Outputs.OpensearchPrivateIps.Value
     case type == CONST_PG:
          nodeIps := infra.Outputs.PostgresqlPrivateIps.Value
     case type == CONST_CS:
	  nodeIps := infra.Outputs.ChefServerPrivateIps.Value
     case type == CONST_A2:
	  nodeIps := infra.Outputs.AutomatePrivateIps.Value
     case type == CONST_FE: 
           nodeIps := append(infra.Outputs.AutomatePrivateIps.Value, 
      infra.Outputs.ChefServerPrivateIps.Value)
	  if ip != "" {
		  if isValidIP(ip, nodeIps) {
			  return []string{ip}, nil
		  }
		  return []string{}, errors.New("Please Enter Valid Node IP")
	  } else if single {
		  ip, err := GetSingleIp(nodeIps)
		  return []string{ip}, err
	  }
	  return nodeIps, nil
}
`

Reply:
Done